recipes/docker-image-guide.md

```diff
@@ -0,0 +1,201 @@
+COPY --chown=node:node package*.json .
+# https://cheatsheetseries.owasp.org/cheatsheets/NodeJS_Docker_Cheat_Sheet.html#2-install-only-production-dependencies-in-the-nodejs-docker-image
+RUN npm ci --only=production
+COPY --chown=node:node . .
+
+FROM install AS run
+# https://cheatsheetseries.owasp.org/cheatsheets/NodeJS_Docker_Cheat_Sheet.html#4-dont-run-containers-as-root
+USER node
+CMD [ "node", "index.js" ]
+```
+
+### Important notes:
+
+1. **Always** specify `.dockerignore` files to reduce security risks and image
+   footprint size. Also, by avoiding sending unwanted files to the builder,
+   build speed is also improved. The file should at least include
+   `node_modules`, `.git` and `.env` files.
+
+2. [The order of Dockerfile instructions matters](https://docs.docker.com/build/guide/layers/).
+
+3. `FROM node:20.8.1-bookworm-slim@sha256:01...53 AS configure`
+
+- Selecting the appropriate Docker image is crucial to achieve minimal resource
+  utilization and minimize vulnerability risks.
+- It is recommended to always use official docker images even though they are not the smallest
+  ones. An excellent illustration of this is the "alpine" image, which, while having
+  a minimal footprint, has experimental support.
+- Include image sha256 hash to ensure that always the same image is downloaded.
+- Use explicit and deterministic Docker base image tags to improve readability
+  and maintainability.
+- Currently, the official `bookworm-slim` image appears to be the most suitable
+  choice for a Node.js image, given its minimal size (nearly equivalent to the
+  Alpine version). Read more about chosing the best NodeJs Docker image
+  [here](https://snyk.io/blog/choosing-the-best-node-js-docker-image).
+
+4. `RUN apt update && apt install dumb-init`
+
+- NodeJs is not designed to be a PID 1 process so we are using a process wrapper
+  to handle termination signals for us instead of doing it manually inside our
+  NodeJs application.
+- Notice how these two commands (apt update and apt install dumb-init) are chained.
+  By doing so we are saving some image footprint size because each docker image step
+  adds an additional layer which affects the final size. That being said, it is
+  recommended whenever is possible to chain RUN commands into single command.
+
+5. `ENTRYPOINT ["dumb-init", "--"]`
+
+- As explained above, we are using dump-init as PID 1 process.
+
+6. `WORKDIR /usr/src/app`
+
+- Application code should be placed inside `/usr/src` subdirectory.
+
+7. `ENV NODE_ENV production`
+
+- If you are building your image for production this ensures that all frameworks
+  and libraries are using the optimal settings for performance and security.
+
+8. `COPY --chown=node:node package*.json .`
+
+- The default Docker behavior is that copied files are owned by `root`.
+  By specifying `--chown=node:node` we are telling that `package\*.json` files
+  will be owned by `node` user instead of `root`.
```

Comment:
Should we specify why is there a need to switch to node user?

- The default Docker behavior is that copied files are owned by `root`.
  By specifying `--chown=node:node` we are telling that `package\*.json` files
  will be owned by `node` user instead of `root`. `node` is the least-privileged user. In case our web application gets compromised, this solution limits number of actions an attacker can do.